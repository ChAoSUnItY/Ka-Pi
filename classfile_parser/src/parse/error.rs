use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match opcode but got opcode value {0}")]
    MatchOutOfBoundOpcode(u8),
    #[error("attempt to match sub opcode in wide opcode but got {0}, while {1:?} is allowed as sub opcode of wide opcode")]
    MatchOutOfBoundWideOpcode(u8, Vec<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
