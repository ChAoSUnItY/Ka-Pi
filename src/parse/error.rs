use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error("internal io error")]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match {0} in {1:?} but got opcode {2}{}", if let Some(opcode) = .3 { format!(" (Which is {:?})", opcode)} else { String::new() })]
    MatchOutOfBoundOpcode(&'static str, Vec<Opcode>, u8, Option<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
