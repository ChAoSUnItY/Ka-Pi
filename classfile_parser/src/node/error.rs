use crate::parse::ParseError;
use thiserror::Error;

pub type NodeResResult<T> = Result<T, NodeResError>;

#[derive(Debug, Error)]
pub enum NodeResError {
    #[error(transparent)]
    ParseFail(#[from] ParseError),
    #[error("failed to parse bytes to string in constant Utf8, given bytes: {0:?}")]
    StringParseFail(Box<[u8]>),
    #[error("unknown constant reference #{0}")]
    UnknownConstantReference(u16),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBound(&'static str, Vec<&'static str>, usize),
    #[error("expected referenced constant {0} at #{1} but got {2}")]
    MismatchReferenceConstant(&'static str, u16, &'static str),
    #[error("character {0} cannot be converted into wildcard")]
    InvalidWildcard(char),
    #[error("character {0} cannot be converted into base type")]
    InvalidBaseType(char),
}
