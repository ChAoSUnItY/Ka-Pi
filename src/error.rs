use std::{
    error::Error,
    fmt::{Debug, Display},
};
use std::fmt::write;

use jni::errors::Error as JniError;

use crate::error::KapiError::*;

pub(crate) trait IntoKapiResult<T> {
    fn into_kapi(self) -> KapiResult<T>;
}

pub type KapiResult<T> = Result<T, KapiError>;

impl<T, E> IntoKapiResult<T> for Result<T, E>
where
    E: Into<KapiError>,
{
    fn into_kapi(self) -> KapiResult<T> {
        self.map_err(|e| e.into())
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum KapiError {
    StateError(&'static str),
    Utf8Error(&'static str),
    TypeError(&'static str),
    ArgError(String),
    ClassResolveError(&'static str),
    ClassParseError(String),
    JNIError(String),
}

impl Display for KapiError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StateError(cause) => write!(f, "{}", cause),
            Utf8Error(cause) => write!(f, "{}", cause),
            TypeError(cause) => write!(f, "{}", cause),
            ArgError(cause) => write!(f, "{}", cause),
            ClassResolveError(cause) => write!(f, "{}", cause),
            ClassParseError(cause) => write!(f, "{}", cause),
            JNIError(cause) => write!(f, "{}", cause),
        }
    }
}

impl Error for KapiError {}

impl From<jni::errors::Error> for KapiError {
    fn from(value: JniError) -> Self {
        JNIError(value.to_string())
    }
}
