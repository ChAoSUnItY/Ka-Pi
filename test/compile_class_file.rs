use insta::assert_yaml_snapshot;

use ka_pi::error::KapiResult;

#[test]
fn test_main() -> KapiResult<()> {
    let class_tree = ka_pi::asm::parse::to_class(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    ))?;

    assert_yaml_snapshot!(class_tree);

    Ok(())
}
