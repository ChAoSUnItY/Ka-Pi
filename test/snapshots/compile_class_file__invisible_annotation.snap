---
source: test/compile_class_file.rs
expression: "to_class(include_bytes!(\"../compiled_source/out/production/compiled_source/InvisibleAnnotation.class\"))?"
---
java_version: V17
constant_pool_count: 43
constant_pool:
  len: 43
  entries:
    1:
      Class:
        name_index: 2
    2:
      Utf8:
        length: 19
        bytes:
          - 73
          - 110
          - 118
          - 105
          - 115
          - 105
          - 98
          - 108
          - 101
          - 65
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
    3:
      Class:
        name_index: 4
    4:
      Utf8:
        length: 16
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 79
          - 98
          - 106
          - 101
          - 99
          - 116
    5:
      Class:
        name_index: 6
    6:
      Utf8:
        length: 31
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 97
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 47
          - 65
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
    7:
      Utf8:
        length: 1
        bytes:
          - 98
    8:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 66
    9:
      Utf8:
        length: 17
        bytes:
          - 65
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 68
          - 101
          - 102
          - 97
          - 117
          - 108
          - 116
    10:
      Integer:
        bytes:
          - 0
          - 0
          - 0
          - 1
    11:
      Utf8:
        length: 1
        bytes:
          - 115
    12:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 83
    13:
      Utf8:
        length: 1
        bytes:
          - 105
    14:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 73
    15:
      Utf8:
        length: 1
        bytes:
          - 102
    16:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 70
    17:
      Float:
        bytes:
          - 63
          - 128
          - 0
          - 0
    18:
      Utf8:
        length: 1
        bytes:
          - 106
    19:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 74
    20:
      Long:
        high_bytes:
          - 0
          - 0
          - 0
          - 0
        low_bytes:
          - 0
          - 0
          - 0
          - 1
    22:
      Utf8:
        length: 1
        bytes:
          - 100
    23:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 68
    24:
      Double:
        high_bytes:
          - 63
          - 240
          - 0
          - 0
        low_bytes:
          - 0
          - 0
          - 0
          - 0
    26:
      Utf8:
        length: 6
        bytes:
          - 115
          - 116
          - 114
          - 105
          - 110
          - 103
    27:
      Utf8:
        length: 20
        bytes:
          - 40
          - 41
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
    28:
      Utf8:
        length: 3
        bytes:
          - 75
          - 69
          - 75
    29:
      Utf8:
        length: 1
        bytes:
          - 101
    30:
      Utf8:
        length: 8
        bytes:
          - 40
          - 41
          - 76
          - 69
          - 110
          - 117
          - 109
          - 59
    31:
      Utf8:
        length: 6
        bytes:
          - 76
          - 69
          - 110
          - 117
          - 109
          - 59
    32:
      Utf8:
        length: 4
        bytes:
          - 75
          - 65
          - 80
          - 73
    33:
      Utf8:
        length: 10
        bytes:
          - 83
          - 111
          - 117
          - 114
          - 99
          - 101
          - 70
          - 105
          - 108
          - 101
    34:
      Utf8:
        length: 24
        bytes:
          - 73
          - 110
          - 118
          - 105
          - 115
          - 105
          - 98
          - 108
          - 101
          - 65
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 46
          - 106
          - 97
          - 118
          - 97
    35:
      Utf8:
        length: 25
        bytes:
          - 82
          - 117
          - 110
          - 116
          - 105
          - 109
          - 101
          - 86
          - 105
          - 115
          - 105
          - 98
          - 108
          - 101
          - 65
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 115
    36:
      Utf8:
        length: 29
        bytes:
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 97
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 47
          - 84
          - 97
          - 114
          - 103
          - 101
          - 116
          - 59
    37:
      Utf8:
        length: 5
        bytes:
          - 118
          - 97
          - 108
          - 117
          - 101
    38:
      Utf8:
        length: 34
        bytes:
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 97
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 47
          - 69
          - 108
          - 101
          - 109
          - 101
          - 110
          - 116
          - 84
          - 121
          - 112
          - 101
          - 59
    39:
      Utf8:
        length: 4
        bytes:
          - 84
          - 89
          - 80
          - 69
    40:
      Utf8:
        length: 32
        bytes:
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 97
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 47
          - 82
          - 101
          - 116
          - 101
          - 110
          - 116
          - 105
          - 111
          - 110
          - 59
    41:
      Utf8:
        length: 38
        bytes:
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 97
          - 110
          - 110
          - 111
          - 116
          - 97
          - 116
          - 105
          - 111
          - 110
          - 47
          - 82
          - 101
          - 116
          - 101
          - 110
          - 116
          - 105
          - 111
          - 110
          - 80
          - 111
          - 108
          - 105
          - 99
          - 121
          - 59
    42:
      Utf8:
        length: 5
        bytes:
          - 67
          - 76
          - 65
          - 83
          - 83
access_flags:
  - Public
  - Interface
  - Abstract
  - Annotation
this_class: 1
super_class: 3
interfaces_count: 1
interfaces:
  - 5
fields_count: 0
fields: []
methods_count: 8
methods:
  - access_flags:
      - Public
      - Abstract
    name_index: 7
    descriptor_index: 8
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 66
          - 0
          - 10
        attribute:
          AnnotationDefault:
            default_value:
              tag: 66
              value:
                ConstValue:
                  const_value_index: 10
  - access_flags:
      - Public
      - Abstract
    name_index: 11
    descriptor_index: 12
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 83
          - 0
          - 10
        attribute:
          AnnotationDefault:
            default_value:
              tag: 83
              value:
                ConstValue:
                  const_value_index: 10
  - access_flags:
      - Public
      - Abstract
    name_index: 13
    descriptor_index: 14
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 73
          - 0
          - 10
        attribute:
          AnnotationDefault:
            default_value:
              tag: 73
              value:
                ConstValue:
                  const_value_index: 10
  - access_flags:
      - Public
      - Abstract
    name_index: 15
    descriptor_index: 16
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 70
          - 0
          - 17
        attribute:
          AnnotationDefault:
            default_value:
              tag: 70
              value:
                ConstValue:
                  const_value_index: 17
  - access_flags:
      - Public
      - Abstract
    name_index: 18
    descriptor_index: 19
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 74
          - 0
          - 20
        attribute:
          AnnotationDefault:
            default_value:
              tag: 74
              value:
                ConstValue:
                  const_value_index: 20
  - access_flags:
      - Public
      - Abstract
    name_index: 22
    descriptor_index: 23
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 68
          - 0
          - 24
        attribute:
          AnnotationDefault:
            default_value:
              tag: 68
              value:
                ConstValue:
                  const_value_index: 24
  - access_flags:
      - Public
      - Abstract
    name_index: 26
    descriptor_index: 27
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 3
        info:
          - 115
          - 0
          - 28
        attribute:
          AnnotationDefault:
            default_value:
              tag: 115
              value:
                ConstValue:
                  const_value_index: 28
  - access_flags:
      - Public
      - Abstract
    name_index: 29
    descriptor_index: 30
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 9
        attribute_len: 5
        info:
          - 101
          - 0
          - 31
          - 0
          - 32
        attribute:
          AnnotationDefault:
            default_value:
              tag: 101
              value:
                EnumConstValue:
                  type_name_index: 31
                  const_name_index: 32
attributes_count: 2
attributes:
  - attribute_name_index: 33
    attribute_len: 2
    info:
      - 0
      - 34
    attribute:
      SourceFile:
        source_file_index: 34
  - attribute_name_index: 35
    attribute_len: 27
    info:
      - 0
      - 2
      - 0
      - 36
      - 0
      - 1
      - 0
      - 37
      - 91
      - 0
      - 1
      - 101
      - 0
      - 38
      - 0
      - 39
      - 0
      - 40
      - 0
      - 1
      - 0
      - 37
      - 101
      - 0
      - 41
      - 0
      - 42
    attribute:
      RuntimeVisibleAnnotations:
        num_annotations: 2
        annotations:
          - type_index: 36
            num_element_value_pairs: 1
            element_value_pairs:
              - element_name_index: 37
                value:
                  tag: 91
                  value:
                    ArrayValue:
                      num_values: 1
                      values:
                        - tag: 101
                          value:
                            EnumConstValue:
                              type_name_index: 38
                              const_name_index: 39
          - type_index: 40
            num_element_value_pairs: 1
            element_value_pairs:
              - element_name_index: 37
                value:
                  tag: 101
                  value:
                    EnumConstValue:
                      type_name_index: 41
                      const_name_index: 42

